### PR TITLE
Method to upload for Mini-Apps

### PR DESCRIPTION
**Pull request by hamiltwp**
I have added an upload method using NIO library, which also implements specified landing zone.

### CODE REVIEW

**Comment by rpastrana** [comment]
@hamiltwp thanks for your submission. In order to get your code merged in, we need to do a formal code review. 
Also, all code checked in needs to be formatted uniformly. If you're using Eclipse to develop, please use this formatter: https://github.com/hpcc-systems/HPCC-JAPIs/blob/master/eclipse/HPCC-JAVA-Formatter.xml

Instructions are at the bottom of the JAPI readme. Thanks.

**Comment by rpastrana** [comment]
@Michael-Gardner please review.

**Comment by hamiltwp** [comment]
public boolean upLoadLandingZone(File uploadFile, DropZone dropZone) is the added method.  The changes in the other methods, and comments, are due to applying HPCC-JAVA-Formatter

**Comment by rpastrana** [comment]
@hamiltwp the formatter caused more issues than it cleaned up. Can you rewind to the previous commit? I'll handle the formatting afterwards. Also my comments will be based on the first commit.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 1]
```diff
@@ -718,7 +722,74 @@ public boolean uploadFileLocalDropZone(File file) throws Exception
 
         return uploadFile(file, fetchLocalDropZones[0]);
     }
+    public boolean upLoadLandingZone(File uploadFile,DropZone dropZone) {
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: It's not immediately obvious what this method does when compared to uploadFile(File file, DropZone dropZone).
Please rename the method so it's self documenting, also a block comment describing the method would help.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 2]
```diff
@@ -718,7 +722,74 @@ public boolean uploadFileLocalDropZone(File file) throws Exception
...
+  	
+    	String boundary = Utils.createBoundary();
+    	
+    	URLConnection fileUploadConnection = null;
+        URL fileUploadURL = null;
+        String FILESPRAYWSDLURI     = "/FileSpray";
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: use the member var FILESPRAYWSDLURI   and UPLOADURI

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 3]
```diff
@@ -718,7 +722,74 @@ public boolean uploadFileLocalDropZone(File file) throws Exception
...
+        String UPLOADURI           = FILESPRAYWSDLURI + "/UploadFile?upload_";
+        String uploadurlbuilder = UPLOADURI;
+        uploadurlbuilder += "&NetAddress=" + dropZone.getNetAddress();
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: we might want to ensure dropzone != Null before this operation

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java (hunk 2 above)
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: We might as well ensure File is valid right up front before setting up the connection

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 4]
```diff
@@ -718,7 +722,74 @@ public boolean uploadFileLocalDropZone(File file) throws Exception
...
+			 
+			   // MappedByteBuffer buffer = inChannel.map(FileChannel.MapMode.READ_ONLY, 0, inChannel.size());
+			WritableByteChannel outchannel = Channels.newChannel(output);
+			while(inChannel.read(buffer) > 0){
+			    buffer.flip();
+			    while(buffer.hasRemaining()) {
+			    	outchannel.write(buffer);
+			    }
+			   // buffer.rewind();
+			   // buffer.clear();
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: let's remove any leftover comments

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 5]
```diff
@@ -718,7 +722,74 @@ public boolean uploadFileLocalDropZone(File file) throws Exception
...
+	        	Utils.startMulti(output, uploadFile.getName(), boundary, "");
+			InputStream input = new FileInputStream(uploadFile.getAbsolutePath());
+			
+			//using channels
+			ByteBuffer buffer = ByteBuffer.allocate(1000000);
+			
+			RandomAccessFile aFile     = new RandomAccessFile(uploadFile.getAbsolutePath(), "rw");
+			    FileChannel      inChannel = aFile.getChannel();
+			 
+			   // MappedByteBuffer buffer = inChannel.map(FileChannel.MapMode.READ_ONLY, 0, inChannel.size());
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: Remove commented out lines

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 6]
```diff
@@ -718,7 +722,74 @@ public boolean uploadFileLocalDropZone(File file) throws Exception
...
+	            return true;
+		} catch (MalformedURLException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		} 
+        
+
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: Close out all streams, files. A finally block is convenient for this.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 7]
```diff
@@ -718,7 +722,74 @@ public boolean uploadFileLocalDropZone(File file) throws Exception
...
+			    buffer.flip();
+			    while(buffer.hasRemaining()) {
+			    	outchannel.write(buffer);
+			    }
+			   // buffer.rewind();
+			   // buffer.clear();
+			}
+			
+		   Utils.closeMulti(output, boundary);
+		   outchannel.close();
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: an exception could prevent this lines from executing, see my comment below about a finally clause

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 8]
```diff
@@ -718,7 +722,74 @@ public boolean uploadFileLocalDropZone(File file) throws Exception
...
+	        	fileUploadConnection.setDoOutput(true);
+	        	fileUploadConnection.setRequestProperty("Content-Type", "multipart/form-data; boundary=" + boundary);
+	        	fileUploadConnection.setRequestProperty("Authorization", fsconn.getBasicAuthString());
+	        	OutputStream output = fileUploadConnection.getOutputStream();
+
+	        	Utils.startMulti(output, uploadFile.getName(), boundary, "");
+			InputStream input = new FileInputStream(uploadFile.getAbsolutePath());
+			
+			//using channels
+			ByteBuffer buffer = ByteBuffer.allocate(1000000);
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: Is this a reasonable buffer size? Also we could make a multiple of the member var BUFFER_LENGTH

**Comment by hamiltwp** [comment]
@rpastrana I have addressed your questions in latest commit.

**Comment by dabayliss** [comment]
@rpastrana - we good now ? we have a hipie PR pending this change

**Comment by rpastrana** [comment]
@dabayliss I'll review the latest changes now. I think your team builds off of our Maven release (not from source) correct? If so we might need to deploy a 1 change release for you guys...

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 9]
```diff
@@ -722,15 +722,25 @@ public boolean uploadFileLocalDropZone(File file) throws Exception
...
+        if (uploadFile == null || dropZone == null){
+            return false;
+        }
         String boundary = Utils.createBoundary();
-
+        Boolean returnValue = true;
         URLConnection fileUploadConnection = null;
         URL fileUploadURL = null;
-        String FILESPRAYWSDLURI = "/FileSpray";
         String UPLOADURI = FILESPRAYWSDLURI + "/UploadFile?upload_";
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: Why're we redefining this here? Can we use the member var on line 131?

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 10]
```diff
@@ -749,27 +759,36 @@ public boolean upLoadLandingZone(File uploadFile, DropZone dropZone)
...
-            inChannel.close();
-            output.close();
-            input.close();
-            aFile.close();
-            // server response
+            catch (IOException e){
+                e.printStackTrace();
+                returnValue = false;
+            }
+            finally {
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: we might be better off moving this finally to the outer try block which contains the entire life-cycle of the buffers/streams that need to be closed. 
BTW, do you need to test if outchannel, inchannel != null and /or if they're open?

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 11]
```diff
@@ -749,27 +759,36 @@ public boolean upLoadLandingZone(File uploadFile, DropZone dropZone)
...
-
-            Utils.closeMulti(output, boundary);
-            outchannel.close();
-            inChannel.close();
-            output.close();
-            input.close();
-            aFile.close();
-            // server response
+            catch (IOException e){
+                e.printStackTrace();
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: the stack trace is ok, but we might be better off exposing the exception message. 
Use the Utils.println mechanism for this.

for example:
 Utils.println(System.err, "Encountered error while reading file: " +  e.getLocalMessage() , false, verbose);

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 12]
```diff
@@ -778,20 +797,21 @@ public boolean upLoadLandingZone(File uploadFile, DropZone dropZone)
             {
                 response.append(line);
             }
-            return true;
         }
         catch (MalformedURLException e)
         {
             // TODO Auto-generated catch block
             e.printStackTrace();
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: See above message regarding useful error message reporting.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 13]
```diff
@@ -778,20 +797,21 @@ public boolean upLoadLandingZone(File uploadFile, DropZone dropZone)
...
+            returnValue = false;
         }
         catch (IOException e)
         {
             // TODO Auto-generated catch block
             e.printStackTrace();
+            returnValue = false;
```
Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: Let's output the exception's message here as well.

**Comment by rpastrana** [comment]
@hamiltwp @dabayliss latest changes reviewed please take a look at comments.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java (hunk 12 above)
Comment (hamiltwp) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: I think this is already addressed in the latest commit

Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: This and the following catch still need to be updated (based on latest commit I see ( c2063cf)

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java [hunk 14]
```diff
@@ -718,7 +722,96 @@ public boolean uploadFileLocalDropZone(File file) throws Exception
...
+    /**
+     * UPLOADS A FILE TO THE SPECIFIED LANDING ZONE
+     * USED BY MINIAPPS FOR UP TO 2GB FILE SIZES
+     * @param file
+     *            - The File to upload
+     * @param dropZone
+     *            - The target dropzone 
+     * @return - Boolean, success
+     */
+    public boolean upLoadMiniAppFile(File uploadFile, DropZone dropZone)
```
Comment (Michael-Gardner) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: Can you please change this to uploadMiniAppFile() so it fits the same camel case that we use for uploadFile?

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java (hunk 10 above)
Comment (hamiltwp) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: in the outer block there may be no channels to close, right?

Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: It depends in which scope you declare the channels, if you need to, they can be declared outside the outer try. 
Bottom line is if line 766 or others throw an exception, the channels are not closed.

**Comment by rpastrana** [comment]
@hamiltwp there are still a few outstanding issues w/ this pull request, please let me know when this is ready to go. thanks.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java (hunk 10 above)
Comment (hamiltwp) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: The final block closes if 766 throws an exception.   Are we wanting a catch and close on opening the channels, to close the channels?

Comment (rpastrana) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCFileSprayClient.java]: you'd prob want to ensure the channel is open before attempting to close it.

**Comment by hamiltwp** [comment]
@rpastrana the channel close is removed to the outer finally block.  Does that cover it?

**Comment by hamiltwp** [comment]
accidentally closed

**Comment by rpastrana** [comment]
@hamiltwp please confirm you addressed the issue you mentioned regarding the state of the channels in the finally block, you might need to ensure the channel is open before closing, unless the close() handles closed channels efficiently and gracefully.
Also, there were other comments - The 2 catch blocks need to be cleaned up, and they should output meaningful information via the Utils.println mechanism. Also I think we have a local var which duplicates a member var.

**Comment by hamiltwp** [comment]
@rpastrana, confirmed:

- [ ]  the state of the channels in the finally block
- [ ] closing channel error is trapped
- [ ] outter catch block prints error for malformed url and file open
- [ ]  local var which duplicates a member var , addressed in prior commit

**Comment by dabayliss** [comment]
excellent work folks. Thank you @rpastrana.

Just my opinion - but on the formatting thing - I **always** insist that formatting changes happen on a different PR to semantic ones. At my age my eyes get sloppy and if you have a couple of semantic tweaks inside a thousand format changes I will miss them ...

**Comment by rpastrana** [comment]
Agreed, in this case it was the new code that needed to be reformatted. My change included changes to legacy code since it was a "cleanup" change.
